[AArch64][GlobalISel] Perfect Shuffles

davemgreen opened the pull request:
This is a port of the existing perfect shuffle generation code from SDAG, geneticized to work for both SDAG and GISel.  I wrote it a while ago and it has been sitting on my machine. It brings the codegen for certain shuffles inline and avoids the need for generating a tbl and constant pool load.

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:
```diff
@@ -504,6 +504,189 @@ void applyINS(MachineInstr &MI, MachineRegisterInfo &MRI,
+      ExtLane = MaskElt < 4 ? MaskElt : (MaskElt - 4);
+      Input = MaskElt < 4 ? V1 : V2;
+      if (VT.getScalarSizeInBits() == 16 && VT != LLT::fixed_vector(4, 16)) {
+        Input = MIB.buildInstr(TargetOpcode::G_BITCAST,
```
buildBitcast

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:
```diff
@@ -504,6 +504,189 @@ void applyINS(MachineInstr &MI, MachineRegisterInfo &MRI,
+        Input = MIB.buildInstr(TargetOpcode::G_BITCAST,
+                               {LLT::fixed_vector(4, 16)}, {Input})
+                    .getReg(0);
+        OpLHS = MIB.buildInstr(TargetOpcode::G_BITCAST,
```
buildBitcast

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:
```diff
@@ -504,6 +504,189 @@ void applyINS(MachineInstr &MI, MachineRegisterInfo &MRI,
+        Input = MIB.buildInstr(TargetOpcode::G_BITCAST,
+                               {LLT::fixed_vector(2, 64)}, {Input})
+                    .getReg(0);
+        OpLHS = MIB.buildInstr(TargetOpcode::G_BITCAST,
```
buildBitcast

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:
```diff
@@ -504,6 +504,189 @@ void applyINS(MachineInstr &MI, MachineRegisterInfo &MRI,
+                    .getReg(0);
+      }
+      if (VT.getScalarSizeInBits() == 32 && VT != LLT::fixed_vector(2, 64)) {
+        Input = MIB.buildInstr(TargetOpcode::G_BITCAST,
```
buildBitcast

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:
```diff
@@ -504,6 +504,189 @@ void applyINS(MachineInstr &MI, MachineRegisterInfo &MRI,
+      ExtLane = MaskElt < 2 ? MaskElt : (MaskElt - 2);
+      Input = MaskElt < 2 ? V1 : V2;
+      if (VT.getScalarSizeInBits() == 16 && VT != LLT::fixed_vector(2, 32)) {
+        Input = MIB.buildInstr(TargetOpcode::G_BITCAST,
```
buildBitcast

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:
```diff
@@ -504,6 +504,189 @@ void applyINS(MachineInstr &MI, MachineRegisterInfo &MRI,
+        Input = MIB.buildInstr(TargetOpcode::G_BITCAST,
+                               {LLT::fixed_vector(2, 32)}, {Input})
+                    .getReg(0);
+        OpLHS = MIB.buildInstr(TargetOpcode::G_BITCAST,
```
buildBitcast

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:
```diff
@@ -504,6 +504,189 @@ void applyINS(MachineInstr &MI, MachineRegisterInfo &MRI,
+    auto Ins = MIB.buildInsertVectorElementConstant(MRI.getType(Input), OpLHS,
+                                                    Ext, RHSID & 0x3);
+    if (MRI.getType(Ins.getReg(0)) != VT)
+      return MIB.buildInstr(TargetOpcode::G_BITCAST, {VT}, {Ins}).getReg(0);
```
buildBitcast

madhur13490 commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:
```diff
@@ -504,6 +504,189 @@ void applyINS(MachineInstr &MI, MachineRegisterInfo &MRI,
+      PFIndexes[i] = ShuffleMask[i];
+  }
+
+  // Compute the index in the perfect shuffle table.
```
A comment explaining constants used for multiplication would be good.

madhur13490 commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:
```diff
@@ -504,6 +504,189 @@ void applyINS(MachineInstr &MI, MachineRegisterInfo &MRI,
+static Register GeneratePerfectShuffle(unsigned ID, Register V1, Register V2,
+                                       unsigned PFEntry, Register LHS,
+                                       Register RHS, MachineIRBuilder &MIB,
+                                       MachineRegisterInfo &MRI) {
```
A comment please explaining the below calculations.

tschuett commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:
```diff
@@ -504,6 +504,189 @@ void applyINS(MachineInstr &MI, MachineRegisterInfo &MRI,
 
+/// Match 4 elemental G_SHUFFLE_VECTOR
+bool matchPerfectShuffle(MachineInstr &MI, MachineRegisterInfo &MRI) {
+  assert(MI.getOpcode() == TargetOpcode::G_SHUFFLE_VECTOR);
```
cast to GShuffleVector

tschuett commented on llvm/lib/Target/AArch64/AArch64PerfectShuffle.h:
```diff
@@ -6588,6 +6588,25 @@ static const unsigned PerfectShuffleTable[6561 + 1] = {
     835584U,     // <u,u,u,u>: Cost 0 copy LHS
     0};
 
+enum {
```
Would `enum class` create too much noise?

aemerson commented:
Brilliant, thanks for doing this.

aemerson commented:
Reverse ping @davemgreen

davemgreen commented:
Hi - I will hopefully get back to addressing this soon - I have been going through some other stuff recently (fp128, some of the missing combines and inefficiencies is how we lower small vectors).

davemgreen commented:
Rebase and ping - thanks.

arsenm commented on llvm/include/llvm/CodeGen/GlobalISel/MachineIRBuilder.h:
```diff
@@ -1365,6 +1365,24 @@ class MachineIRBuilder {
+                                                       const SrcOp &Val,
+                                                       const SrcOp &Elt,
+                                                       const int Idx) {
+    auto TLI = getMF().getSubtarget().getTargetLowering();
```
no auto

arsenm commented on llvm/include/llvm/CodeGen/GlobalISel/MachineIRBuilder.h:
```diff
@@ -1365,6 +1365,24 @@ class MachineIRBuilder {
+                                                       const SrcOp &Elt,
+                                                       const int Idx) {
+    auto TLI = getMF().getSubtarget().getTargetLowering();
+    unsigned VecIdxWidth = TLI->getVectorIdxTy(getDataLayout()).getSizeInBits();
```
Should have a version of this that directly returns LLT. 

I'm also not sure how I feel about requiring TLI in these low level build functions. For the most part, you could just use i64 and let it legalize as required (but this does present an issue after legalization)

davemgreen committed:
[AArch64][GlobalISel] Perfect Shuffles

This is a port of the existing perfect shuffle generation code from SDAG,
geneticized to work for both SDAG and GISel.  I wrote it a while ago and it has
been sitting on my machine. It brings the codegen for certain shuffles inline
and avoids the need for generating a tbl and constant pool load.

davemgreen commented:
I've rebased this over the other commit that was included in it - Thanks.

arsenm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -14090,8 +13924,95 @@ SDValue AArch64TargetLowering::LowerVECTOR_SHUFFLE(SDValue Op,
+        ExtSrc = DAG.getBitcast(MVT::v4f16, ExtSrc);
+        InsSrc = DAG.getBitcast(MVT::v4f16, InsSrc);
+      }
+      SDValue Ext = DAG.getNode(ISD::EXTRACT_VECTOR_ELT, dl,
+                                ExtSrc.getValueType().getVectorElementType(),
+                                ExtSrc, DAG.getVectorIdxConstant(ExtLane, dl));
+      SDValue Ins =
+          DAG.getNode(ISD::INSERT_VECTOR_ELT, dl, ExtSrc.getValueType(), InsSrc,
+                      Ext, DAG.getVectorIdxConstant(InsLane, dl));
```
can query the index type once instead of repeating it for each of these cases

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:
```diff
@@ -524,6 +524,129 @@ void applyINS(MachineInstr &MI, MachineRegisterInfo &MRI,
+      ExtSrc = MIB.buildBitcast(LLT::fixed_vector(2, 64), ExtSrc).getReg(0);
+      InsSrc = MIB.buildBitcast(LLT::fixed_vector(2, 64), InsSrc).getReg(0);
+    }
+    auto Ext = MIB.buildExtractVectorElementConstant(
+        MRI.getType(ExtSrc).getElementType(), ExtSrc, ExtLane);
+    auto Ins = MIB.buildInsertVectorElementConstant(MRI.getType(ExtSrc), InsSrc,
+                                                    Ext, InsLane);
```
ditto

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:
```diff
@@ -524,6 +524,129 @@ void applyINS(MachineInstr &MI, MachineRegisterInfo &MRI,
+    LLT Ty = MRI.getType(OpLHS);
+    switch (OpNum) {
+    default:
+      llvm_unreachable("Unexpected perfect shuffle opcode\n");
```
```suggestion
      llvm_unreachable("Unexpected perfect shuffle opcode");
```

davemgreen committed:
Just use buildConstant/getConstant, we know the type already

arsenm commented on llvm/lib/Target/AArch64/AArch64PerfectShuffle.h:
```diff
@@ -6723,6 +6742,110 @@ inline bool isREVMask(ArrayRef<int> M, unsigned EltSize, unsigned NumElts,
+      ExtLane = MaskElt < 2 ? MaskElt : (MaskElt - 2);
+      Input = MaskElt < 2 ? V1 : V2;
+      return BuildExtractInsert64(Input, ExtLane, OpLHS, RHSID & 0x3);
+    } else {
```
No else after return

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:
```diff
@@ -524,6 +524,133 @@ void applyINS(MachineInstr &MI, MachineRegisterInfo &MRI,
+    switch (OpNum) {
+    default:
+      llvm_unreachable("Unexpected perfect shuffle opcode");
+    case OP_VUZPL:
+      return MIB.buildInstr(AArch64::G_UZP1, {Ty}, {OpLHS, OpRHS}).getReg(0);
+    case OP_VUZPR:
+      return MIB.buildInstr(AArch64::G_UZP2, {Ty}, {OpLHS, OpRHS}).getReg(0);
+    case OP_VZIPL:
+      return MIB.buildInstr(AArch64::G_ZIP1, {Ty}, {OpLHS, OpRHS}).getReg(0);
+    case OP_VZIPR:
+      return MIB.buildInstr(AArch64::G_ZIP2, {Ty}, {OpLHS, OpRHS}).getReg(0);
+    case OP_VTRNL:
+      return MIB.buildInstr(AArch64::G_TRN1, {Ty}, {OpLHS, OpRHS}).getReg(0);
+    case OP_VTRNR:
+      return MIB.buildInstr(AArch64::G_TRN2, {Ty}, {OpLHS, OpRHS}).getReg(0);
```
Can do this one one buildInstr + a utility function to convert from OpNum -> opcode

davemgreen committed:
Address more comments

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:
```diff
@@ -524,6 +524,140 @@ void applyINS(MachineInstr &MI, MachineRegisterInfo &MRI,
+      InsSrc = MIB.buildBitcast(LLT::fixed_vector(2, 32), InsSrc).getReg(0);
+    }
+    auto Ext = MIB.buildExtractVectorElement(
+        MRI.getType(ExtSrc).getElementType(), ExtSrc,
```
Avoid repeated getTypes

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerLowering.cpp:
```diff
@@ -524,6 +524,140 @@ void applyINS(MachineInstr &MI, MachineRegisterInfo &MRI,
+                                   Register OpRHS) {
+    LLT Ty = MRI.getType(OpLHS);
+    unsigned Opc = 0;
+    switch (OpNum) {
+    default:
+      llvm_unreachable("Unexpected perfect shuffle opcode");
+    case OP_VUZPL:
+      Opc = AArch64::G_UZP1;
+      break;
+    case OP_VUZPR:
+      Opc = AArch64::G_UZP2;
+      break;
+    case OP_VZIPL:
+      Opc = AArch64::G_ZIP1;
+      break;
+    case OP_VZIPR:
+      Opc = AArch64::G_ZIP2;
+      break;
+    case OP_VTRNL:
+      Opc = AArch64::G_TRN1;
+      break;
+    case OP_VTRNR:
+      Opc = AArch64::G_TRN2;
+    }
```
Move to standalone utility function with return instead of conditional assign and break

arsenm reviewed (commented):
I know nothing of aarch64 so I just have more nits
